### DISCUSSION TITLE
Building new object with emqx

### DISCUSSION THREAD

**Discussion by olabr** [emqx/emqx#14132] [Q&A] [answered]
Hello,
I'm totally new to emqx and might be a dump question, but I'm looking for best practice to solve such case.

I have a raw input from my industrial environment in a large array of data looking something  like this:

{"timestamp":1730538138601,"values":
[
{"id":"Simulation Examples.Functions.Random1","v":47,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp1","v":83,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp2","v":192.75,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp3","v":40,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp4","v":980,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp5","v":980500,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp6","v":12500,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp7","v":999950005,"q":true,"t":1730538138598},
{"id":"Simulation Examples.Functions.Ramp8","v":198.25,"q":true,"t":1730538138598}
]
}

I would like to map this into a new object to send back to the broker in a structure like this:
{
instancename : statictext,
ts : payload.timestamp,
temperature : <== the .v from .id equal 'Examples.Functions.Ramp5' (only if .q is true),
pressure : <== the .v from .id equal 'Examples.Functions.Ramp8' (only if .q is true),
conductivity : <== the .v from id equal 'Examples.Functions.Ramp6' (only if .q is true)
}

I have no control of the order of the array.

Can anyone please tell me how to archieve this?

Kind regards

**Comment by zmstone** [comment, marked as answer]
```
SELECT
  jq('{
  instancename: "statictext",
  ts: .timestamp,
  temperature: (.values[] | select(.id == "Simulation Examples.Functions.Ramp5" and .q == true) | .v),
  pressure: (.values[] | select(.id == "Simulation Examples.Functions.Ramp8" and .q == true) | .v),
  conductivity: (.values[] | select(.id == "Simulation Examples.Functions.Ramp6" and .q == true) | .v)
}',payload) as output
FROM
  "t/#"
```

output:

```
{
  "output": [
    {
      "conductivity": 12500,
      "instancename": "statictext",
      "pressure": 198.25,
      "temperature": 980500,
      "ts": 1730538138601
    }
  ]
}
```

jq function ref: https://docs.emqx.com/en/emqx/latest/data-integration/rule-sql-jq.html

the output is an array.
if you want `output` to be an object, use `nth(1,jq(...,paylaod))`

**Comment by olabr** [comment]
Thank you @zmstone .

Looks like I have to sit down and teach myself jq :)

Kind regards
